File: mgyoutube-multitech/api-node/src/webservices/supportwebservice.ts
import { ModuleRepoRegistry } from './modulereporegistry';
import { Helpers } from './helpers';
import { Request, Response, Next, Server, createServer } from 'restify'
import { NotFoundError, UnauthorizedError, InternalServerError } from 'restify-errors'
import { ParentsWebService } from './parentswebservice';
import { UseNext } from './whatsnext';


export class SupportWebService {

    public static setupRouter(server: Server) {
        // DRP: restify can't seem to use class methods for handlers here... they lose the "this" pointer reference
        server.put('/api/support/users/:username', async (req, res, next) => {
            return UseNext.handleAsyncRestifyCall(await SupportWebService.createUpdateUserByUsername(req, res), next)
        });
        server.get('/api/support/users/:username', async (req, res, next) => {
            return UseNext.handleAsyncRestifyCall(await SupportWebService.getUserByUsername(req, res), next)
        });
        server.del('/api/support/users/:username', async (req, res, next) => {
            return UseNext.handleAsyncRestifyCall(await SupportWebService.removeUserByUsername(req, res), next)
        });
    }

    private static async createUpdateUserByUsername(req: Request, res: Response): Promise<UseNext> {
        const userJson = req.body
        console.log("createUpdateUserByUsername: userJson=", userJson)

        const sanitizedUserJson = userJson
        const user = Helpers.marshalUserFromJson(sanitizedUserJson)
        console.log("createUpdateUserByUsername: user=", user)
        const createdUser = await ModuleRepoRegistry.getUserModule().createUpdateUser(user)
        console.log("createUpdateUserByUsername: createdUser=", createdUser)
        if (createdUser === undefined) {
            console.error("createUpdateUserByUsername failed for " + sanitizedUserJson);
            const err500 = new InternalServerError();
            return new UseNext(err500);
        }
        createdUser.password = undefined;

        console.log("createUpdateUserByUsername: userJson=", userJson, " returning OK")
        res.setHeader("content-type", "application/json")
        res.send(createdUser)
        return UseNext.Nothing;
    }

    private static async getUserByUsername(req: Request, res: Response): Promise<UseNext> {

        const username = req.params.username
        console.log("getUserByUsername: username=", username)

        const santizedUsername = username
        const user = await ModuleRepoRegistry.getUserModule().getUser(santizedUsername)
        console.log("getUserByUsername: user=", user)

        if (user == null) {
            console.warn("getUserByUsername: username=", username, " returning NOT_FOUND")
            const err404 = new NotFoundError();
            return new UseNext(err404);
        }

        user.password = undefined

        console.log("getUserByUsername: username=", username, " returning OK")

        res.setHeader("content-type", "application/json")
        res.send(user)

        return UseNext.Nothing;
    }

    private static async removeUserByUsername(req: Request, res: Response): Promise<UseNext> {
        const username = req.params.username
        console.log("removeUserByUsername: username=" + username);

        // TODO: need to sanitize payload input before using
        const santizedUsername = username;
        const oldUser = await ModuleRepoRegistry.getUserModule().removeUser(santizedUsername);

        if (oldUser === undefined) {
            console.warn("deleteUserByUsername: username=", username, " returning NOT_FOUND");
            const err404 = new NotFoundError();
            return new UseNext(err404);
        }

        res.send(200)
        return UseNext.Nothing;
    }

}

File: mgyoutube-multitech/api-python/webservices/SupportWebService.py
import json
from pyramid.response import Response
from pyramid.request import Request
from pyramid.view import view_config
from pyramid.config import Configurator
from pyramid.httpexceptions import HTTPNotFound, HTTPUnauthorized, HTTPMethodNotAllowed,  HTTPServerError
from webservices.Helpers import Helpers
from webservices.ModuleRepoRegistry import ModuleRepoRegistry


@view_config(route_name='support-getcreateupatedeleteuser', request_method='GET')
def support_getuser(request):
    username = request.matchdict['username']
    santizedUsername = username
    user = ModuleRepoRegistry.getUserModule().getUser(santizedUsername)
    print("support_getuser: user=", user)

    if (user is None):
        print("support_getuser: username=", username, " returning NOT_FOUND")
        raise HTTPNotFound()

    user.password = None

    responseJson = json.dumps(user.__dict__)

    print("support_getuser: username=", username, " returning OK")
    return Response(body=responseJson, content_type='application/json', charset="UTF-8")


@view_config(route_name='support-getcreateupatedeleteuser', request_method='DELETE')
def support_deleteuser(request):
    username = request.matchdict['username']
    santizedUsername = username
    oldUser = ModuleRepoRegistry.getUserModule().removeUser(santizedUsername)
    print("support_deleteuser: oldUser=", oldUser)

    if (oldUser is None):
        print("support_deleteuser: username=",
              username, " returning NOT_FOUND")
        raise HTTPNotFound()

    print("support_deleteuser: username=", username, " returning OK")
    return Response(status=200)


@view_config(route_name='support-getcreateupatedeleteuser', request_method='PUT')
def support_createuser(request):
    userJson = request.body
    print("createUser: userJson=", userJson)
    sanitizedUserJson = userJson
    user = Helpers.marshalUserFromJson(sanitizedUserJson)
    print("createUser: user=", user)
    createdUser = ModuleRepoRegistry.getUserModule().createUser(user)
    createdUser.password = None

    responseJson = json.dumps(createdUser.__dict__)
    print("createUser: userJson=", userJson,
          " returning OK and ", responseJson)
    return Response(body=responseJson, content_type='application/json', charset="UTF-8")


def setupSupportWebService(config: Configurator):
    config.add_route('support-getcreateupatedeleteuser',
                     '/api/support/users/{username}')
